File: tests/sdk/go/eventbridge_rolearn_test.go
package sdk_test

import (
	"context"
	"fmt"
	"strings"
	"testing"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/eventbridge"
	ebtypes "github.com/aws/aws-sdk-go-v2/service/eventbridge/types"
	"github.com/aws/aws-sdk-go-v2/service/sqs"
	"github.com/aws/aws-sdk-go-v2/service/sts"
)

const ebRoleArnLessSource = "noctaxris.sdk.eb.rolearn"

func TestEventBridgeDeliverSQSWithoutRoleArn(t *testing.T) {
	requireReady(t)
	cfg := loadAWSConfig(t)
	ebClient := newEvents(t, cfg)
	sqsClient := newSQS(t, cfg)
	stsClient := newSTS(t, cfg)
	ctx := context.Background()
	prefix := uniquePrefix(t)

	caller, err := stsClient.GetCallerIdentity(ctx, &sts.GetCallerIdentityInput{})
	if err != nil {
		t.Fatalf("GetCallerIdentity: %v", err)
	}
	accountID := aws.ToString(caller.Account)
	if accountID == "" {
		t.Fatal("GetCallerIdentity missing Account")
	}

	busName := prefix + "-bus"
	ruleName := prefix + "-rule"
	qName := prefix + "-eb-q"

	qOut, err := sqsClient.CreateQueue(ctx, &sqs.CreateQueueInput{QueueName: aws.String(qName)})
	if err != nil {
		t.Fatalf("CreateQueue: %v", err)
	}
	qURL := aws.ToString(qOut.QueueUrl)
	t.Cleanup(func() {
		_, _ = sqsClient.DeleteQueue(ctx, &sqs.DeleteQueueInput{QueueUrl: aws.String(qURL)})
	})
	qARN := queueARN(t, ctx, sqsClient, qURL)

	_, err = ebClient.CreateEventBus(ctx, &eventbridge.CreateEventBusInput{Name: aws.String(busName)})
	if err != nil {
		t.Fatalf("CreateEventBus: %v", err)
	}
	t.Cleanup(func() {
		_, _ = ebClient.DeleteEventBus(ctx, &eventbridge.DeleteEventBusInput{Name: aws.String(busName)})
	})

	pattern := fmt.Sprintf(`{"source":[%q]}`, ebRoleArnLessSource)
	_, err = ebClient.PutRule(ctx, &eventbridge.PutRuleInput{
		Name:         aws.String(ruleName),
		EventBusName: aws.String(busName),
		EventPattern: aws.String(pattern),
		State:        ebtypes.RuleStateEnabled,
	})
	if err != nil {
		t.Fatalf("PutRule: %v", err)
	}
	t.Cleanup(func() {
		_, _ = ebClient.RemoveTargets(ctx, &eventbridge.RemoveTargetsInput{
			Rule:         aws.String(ruleName),
			EventBusName: aws.String(busName),
			Ids:          []string{"sqs"},
		})
		_, _ = ebClient.DeleteRule(ctx, &eventbridge.DeleteRuleInput{
			Name:         aws.String(ruleName),
			EventBusName: aws.String(busName),
		})
	})

	policy := fmt.Sprintf(
		`{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Principal":{"Service":"events.amazonaws.com"},"Action":"sqs:SendMessage","Resource":"%s","Condition":{"ArnLike":{"aws:SourceArn":"arn:aws:events:us-east-1:%s:rule/%s/%s"}}}]}`,
		qARN, accountID, busName, ruleName,
	)
	_, err = sqsClient.SetQueueAttributes(ctx, &sqs.SetQueueAttributesInput{
		QueueUrl:   aws.String(qURL),
		Attributes: map[string]string{"Policy": policy},
	})
	if err != nil {
		t.Fatalf("SetQueueAttributes Policy: %v", err)
	}

	putTargets, err := ebClient.PutTargets(ctx, &eventbridge.PutTargetsInput{
		Rule:         aws.String(ruleName),
		EventBusName: aws.String(busName),
		Targets: []ebtypes.Target{
			{Id: aws.String("sqs"), Arn: aws.String(qARN)},
		},
	})
	if err != nil {
		t.Fatalf("PutTargets: %v", err)
	}
	if putTargets.FailedEntryCount > 0 {
		t.Fatalf("PutTargets FailedEntryCount=%d entries=%+v", putTargets.FailedEntryCount, putTargets.FailedEntries)
	}

	drainQueue(t, ctx, sqsClient, qURL)
	marker := "eb-rolearn-" + prefix
	detail := fmt.Sprintf(`{"marker":%q}`, marker)
	putOut, err := ebClient.PutEvents(ctx, &eventbridge.PutEventsInput{
		Entries: []ebtypes.PutEventsRequestEntry{{
			EventBusName: aws.String(busName),
			Source:       aws.String(ebRoleArnLessSource),
			DetailType:   aws.String("RoleArnLessDelivery"),
			Detail:       aws.String(detail),
		}},
	})
	if err != nil {
		t.Fatalf("PutEvents: %v", err)
	}
	if putOut.FailedEntryCount > 0 {
		t.Fatalf("PutEvents FailedEntryCount=%d entries=%+v", putOut.FailedEntryCount, putOut.Entries)
	}

	body := receiveOneBody(t, ctx, sqsClient, qURL, 8*time.Second)
	if !strings.Contains(body, ebRoleArnLessSource) || !strings.Contains(body, marker) {
		t.Fatalf("SQS body missing delivery markers: %s", body)
	}
}

func TestEventBridgeSkipSQSWithoutRoleArnOrPolicy(t *testing.T) {
	requireReady(t)
	cfg := loadAWSConfig(t)
	ebClient := newEvents(t, cfg)
	sqsClient := newSQS(t, cfg)
	ctx := context.Background()
	prefix := uniquePrefix(t)

	busName := prefix + "-bus"
	ruleName := prefix + "-rule"
	qName := prefix + "-deny-q"

	qOut, err := sqsClient.CreateQueue(ctx, &sqs.CreateQueueInput{QueueName: aws.String(qName)})
	if err != nil {
		t.Fatalf("CreateQueue: %v", err)
	}
	qURL := aws.ToString(qOut.QueueUrl)
	t.Cleanup(func() {
		_, _ = sqsClient.DeleteQueue(ctx, &sqs.DeleteQueueInput{QueueUrl: aws.String(qURL)})
	})
	qARN := queueARN(t, ctx, sqsClient, qURL)

	_, err = ebClient.CreateEventBus(ctx, &eventbridge.CreateEventBusInput{Name: aws.String(busName)})
	if err != nil {
		t.Fatalf("CreateEventBus: %v", err)
	}
	t.Cleanup(func() {
		_, _ = ebClient.DeleteEventBus(ctx, &eventbridge.DeleteEventBusInput{Name: aws.String(busName)})
	})

	pattern := fmt.Sprintf(`{"source":[%q]}`, ebRoleArnLessSource)
	_, err = ebClient.PutRule(ctx, &eventbridge.PutRuleInput{
		Name:         aws.String(ruleName),
		EventBusName: aws.String(busName),
		EventPattern: aws.String(pattern),
		State:        ebtypes.RuleStateEnabled,
	})
	if err != nil {
		t.Fatalf("PutRule: %v", err)
	}
	t.Cleanup(func() {
		_, _ = ebClient.RemoveTargets(ctx, &eventbridge.RemoveTargetsInput{
			Rule:         aws.String(ruleName),
			EventBusName: aws.String(busName),
			Ids:          []string{"deny"},
		})
		_, _ = ebClient.DeleteRule(ctx, &eventbridge.DeleteRuleInput{
			Name:         aws.String(ruleName),
			EventBusName: aws.String(busName),
		})
	})

	putTargets, err := ebClient.PutTargets(ctx, &eventbridge.PutTargetsInput{
		Rule:         aws.String(ruleName),
		EventBusName: aws.String(busName),
		Targets: []ebtypes.Target{
			{Id: aws.String("deny"), Arn: aws.String(qARN)},
		},
	})
	if err != nil {
		t.Fatalf("PutTargets: %v", err)
	}
	if putTargets.FailedEntryCount > 0 {
		t.Fatalf("PutTargets FailedEntryCount=%d entries=%+v", putTargets.FailedEntryCount, putTargets.FailedEntries)
	}

	drainQueue(t, ctx, sqsClient, qURL)
	_, err = ebClient.PutEvents(ctx, &eventbridge.PutEventsInput{
		Entries: []ebtypes.PutEventsRequestEntry{{
			EventBusName: aws.String(busName),
			Source:       aws.String(ebRoleArnLessSource),
			DetailType:   aws.String("RoleArnLessSkip"),
			Detail:       aws.String(`{"probe":"no-policy"}`),
		}},
	})
	if err != nil {
		t.Fatalf("PutEvents: %v", err)
	}
	if msg := receiveOptionalBody(t, ctx, sqsClient, qURL, 2*time.Second); msg != "" {
		t.Fatalf("queue without events.amazonaws.com policy unexpectedly received: %s", msg)
	}
}

File: tests/sdk/nodejs/test/eventbridge-rolearn.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  CreateEventBusCommand,
  DeleteEventBusCommand,
  DeleteRuleCommand,
  PutEventsCommand,
  PutRuleCommand,
  PutTargetsCommand,
  RemoveTargetsCommand,
} from "@aws-sdk/client-eventbridge";
import {
  CreateQueueCommand,
  DeleteMessageCommand,
  DeleteQueueCommand,
  GetQueueAttributesCommand,
  ReceiveMessageCommand,
  SetQueueAttributesCommand,
} from "@aws-sdk/client-sqs";
import { GetCallerIdentityCommand } from "@aws-sdk/client-sts";
import {
  newEvents,
  newSQS,
  newSTS,
  requireReady,
  uniquePrefix,
} from "../lib/helpers.mjs";

const EB_SOURCE = "noctaxris.sdk.eb.rolearn";

async function queueArn(sqs, queueUrl) {
  const attrs = await sqs.send(
    new GetQueueAttributesCommand({
      QueueUrl: queueUrl,
      AttributeNames: ["QueueArn"],
    }),
  );
  const arn = attrs.Attributes?.QueueArn;
  assert.ok(arn, "GetQueueAttributes missing QueueArn");
  return arn;
}

async function drainQueue(sqs, queueUrl) {
  const deadline = Date.now() + 3000;
  while (Date.now() < deadline) {
    const recv = await sqs.send(
      new ReceiveMessageCommand({
        QueueUrl: queueUrl,
        MaxNumberOfMessages: 10,
        WaitTimeSeconds: 1,
        VisibilityTimeout: 0,
      }),
    );
    if (!recv.Messages?.length) return;
    for (const m of recv.Messages) {
      if (!m.ReceiptHandle) continue;
      await sqs.send(
        new DeleteMessageCommand({
          QueueUrl: queueUrl,
          ReceiptHandle: m.ReceiptHandle,
        }),
      );
    }
  }
}

async function receiveOneBody(sqs, queueUrl, timeoutMs) {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    const recv = await sqs.send(
      new ReceiveMessageCommand({
        QueueUrl: queueUrl,
        MaxNumberOfMessages: 1,
        WaitTimeSeconds: 1,
      }),
    );
    if (!recv.Messages?.length) continue;
    const body = recv.Messages[0].Body || "";
    if (recv.Messages[0].ReceiptHandle) {
      await sqs.send(
        new DeleteMessageCommand({
          QueueUrl: queueUrl,
          ReceiptHandle: recv.Messages[0].ReceiptHandle,
        }),
      );
    }
    return body;
  }
  throw new Error(`ReceiveMessage timed out on ${queueUrl}`);
}

async function receiveOptionalBody(sqs, queueUrl, timeoutMs) {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    const recv = await sqs.send(
      new ReceiveMessageCommand({
        QueueUrl: queueUrl,
        MaxNumberOfMessages: 1,
        WaitTimeSeconds: 1,
      }),
    );
    if (!recv.Messages?.length) continue;
    const body = recv.Messages[0].Body || "";
    if (recv.Messages[0].ReceiptHandle) {
      await sqs.send(
        new DeleteMessageCommand({
          QueueUrl: queueUrl,
          ReceiptHandle: recv.Messages[0].ReceiptHandle,
        }),
      );
    }
    return body;
  }
  return "";
}

test("EventBridge deliver SQS without RoleArn", async (t) => {
  if (!(await requireReady(t))) return;
  const eb = newEvents();
  const sqs = newSQS();
  const sts = newSTS();
  const prefix = uniquePrefix();

  const caller = await sts.send(new GetCallerIdentityCommand({}));
  const accountID = caller.Account;
  assert.ok(accountID, "GetCallerIdentity missing Account");

  const busName = `${prefix}-bus`;
  const ruleName = `${prefix}-rule`;
  const qName = `${prefix}-eb-q`;

  const qOut = await sqs.send(new CreateQueueCommand({ QueueName: qName }));
  const qURL = qOut.QueueUrl;
  assert.ok(qURL);
  t.after(async () => {
    try {
      await sqs.send(new DeleteQueueCommand({ QueueUrl: qURL }));
    } catch {
      /* ignore */
    }
  });
  const qARN = await queueArn(sqs, qURL);

  await eb.send(new CreateEventBusCommand({ Name: busName }));
  t.after(async () => {
    try {
      await eb.send(new DeleteEventBusCommand({ Name: busName }));
    } catch {
      /* ignore */
    }
  });

  const pattern = JSON.stringify({ source: [EB_SOURCE] });
  await eb.send(
    new PutRuleCommand({
      Name: ruleName,
      EventBusName: busName,
      EventPattern: pattern,
      State: "ENABLED",
    }),
  );
  t.after(async () => {
    try {
      await eb.send(
        new RemoveTargetsCommand({
          Rule: ruleName,
          EventBusName: busName,
          Ids: ["sqs"],
        }),
      );
    } catch {
      /* ignore */
    }
    try {
      await eb.send(
        new DeleteRuleCommand({ Name: ruleName, EventBusName: busName }),
      );
    } catch {
      /* ignore */
    }
  });

  const policy = JSON.stringify({
    Version: "2012-10-17",
    Statement: [
      {
        Effect: "Allow",
        Principal: { Service: "events.amazonaws.com" },
        Action: "sqs:SendMessage",
        Resource: qARN,
        Condition: {
          ArnLike: {
            "aws:SourceArn": `arn:aws:events:us-east-1:${accountID}:rule/${busName}/${ruleName}`,
          },
        },
      },
    ],
  });
  await sqs.send(
    new SetQueueAttributesCommand({
      QueueUrl: qURL,
      Attributes: { Policy: policy },
    }),
  );

  const putTargets = await eb.send(
    new PutTargetsCommand({
      Rule: ruleName,
      EventBusName: busName,
      Targets: [{ Id: "sqs", Arn: qARN }],
    }),
  );
  assert.equal(putTargets.FailedEntryCount || 0, 0);

  await drainQueue(sqs, qURL);
  const marker = `eb-rolearn-${prefix}`;
  const putOut = await eb.send(
    new PutEventsCommand({
      Entries: [
        {
          EventBusName: busName,
          Source: EB_SOURCE,
          DetailType: "RoleArnLessDelivery",
          Detail: JSON.stringify({ marker }),
        },
      ],
    }),
  );
  assert.equal(putOut.FailedEntryCount || 0, 0);

  const body = await receiveOneBody(sqs, qURL, 8000);
  assert.ok(body.includes(EB_SOURCE), `missing source in ${body}`);
  assert.ok(body.includes(marker), `missing marker in ${body}`);
});

test("EventBridge skip SQS without RoleArn or policy", async (t) => {
  if (!(await requireReady(t))) return;
  const eb = newEvents();
  const sqs = newSQS();
  const prefix = uniquePrefix();

  const busName = `${prefix}-bus`;
  const ruleName = `${prefix}-rule`;
  const qName = `${prefix}-deny-q`;

  const qOut = await sqs.send(new CreateQueueCommand({ QueueName: qName }));
  const qURL = qOut.QueueUrl;
  assert.ok(qURL);
  t.after(async () => {
    try {
      await sqs.send(new DeleteQueueCommand({ QueueUrl: qURL }));
    } catch {
      /* ignore */
    }
  });
  const qARN = await queueArn(sqs, qURL);

  await eb.send(new CreateEventBusCommand({ Name: busName }));
  t.after(async () => {
    try {
      await eb.send(new DeleteEventBusCommand({ Name: busName }));
    } catch {
      /* ignore */
    }
  });

  await eb.send(
    new PutRuleCommand({
      Name: ruleName,
      EventBusName: busName,
      EventPattern: JSON.stringify({ source: [EB_SOURCE] }),
      State: "ENABLED",
    }),
  );
  t.after(async () => {
    try {
      await eb.send(
        new RemoveTargetsCommand({
          Rule: ruleName,
          EventBusName: busName,
          Ids: ["deny"],
        }),
      );
    } catch {
      /* ignore */
    }
    try {
      await eb.send(
        new DeleteRuleCommand({ Name: ruleName, EventBusName: busName }),
      );
    } catch {
      /* ignore */
    }
  });

  const putTargets = await eb.send(
    new PutTargetsCommand({
      Rule: ruleName,
      EventBusName: busName,
      Targets: [{ Id: "deny", Arn: qARN }],
    }),
  );
  assert.equal(putTargets.FailedEntryCount || 0, 0);

  await drainQueue(sqs, qURL);
  await eb.send(
    new PutEventsCommand({
      Entries: [
        {
          EventBusName: busName,
          Source: EB_SOURCE,
          DetailType: "RoleArnLessSkip",
          Detail: JSON.stringify({ probe: "no-policy" }),
        },
      ],
    }),
  );
  const msg = await receiveOptionalBody(sqs, qURL, 2000);
  assert.equal(msg, "", `queue without policy unexpectedly received: ${msg}`);
});
